[PT-D][Sharding] Enable ops needed in the transformer model training

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75374

From the code base of FairSeq and MetaSeq codebase (which is essentially a transformer model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example:

Ops include: chuck, transpose, view, mask_fill, dropout, softmax and type_as.

Isolate the common logic of registering simple ops into a function and for future register, we just need to implement at most three functions for a new op.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

Commit:
[PT-D][Sharding] Enable ops needed in the transformer model training

From the code base of TextRay (which is essentially XLMR model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of TextRay (which is essentially XLMR model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of TextRay (which is essentially XLMR model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of TextRay (which is essentially XLMR model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of TextRay (which is essentially XLMR model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/ops/test_chunk.py:
```diff
@@ -0,0 +1,53 @@
+    @requires_nccl()
+    def test_sharded_chunk(self):
+        specs = generate_chunk_sharding_specs_for_test(0)
+        st = sharded_tensor.rand(specs[0], 12, 16)
+        st.contiguous().view(-1, 4, 4).transpose(1, 2).view(12, 16).contiguous()
+        st = st * 0.2
+        local_tensor = st.local_tensor()
+        local_tensor_chunked = torch.chunk(local_tensor, 4, dim=-1)
```
Maybe have a list of combinations for sharded_tensor size and chunk size to ensure we have enough coverage across all edge cases.

fduwjj replied:
sure.

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/ops/test_matrix_ops.py:
```diff
@@ -0,0 +1,66 @@
+    @skip_if_lt_x_gpu(TEST_GPU_NUM)
+    @requires_nccl()
+    def test_sharded_matrix(self):
+        specs = generate_chunk_sharding_specs_for_test(0)
+        st = sharded_tensor.rand(specs[0], 16, 5)
+        st_2 = sharded_tensor.rand(specs[0], 16, 5)
+        st = st.view(-1, 4, 5).contiguous()
+        st_2 = st_2.view(-1, 4, 5).contiguous()
+        st_3 = torch.bmm(st, st_2.transpose(1, 2))
+        self.assertEqual(st_3.size(), torch.Size([4, 4, 4]))
+        new_spec = copy.deepcopy(specs[0])
+        new_spec.dim = -1
+        st_3 = sharded_tensor.rand(new_spec, 3, 4, 11)
+        st_4 = sharded_tensor.rand(new_spec, 3, 4, 11)
+        st_5 = torch.bmm(st_3, st_4.transpose(1, 2))
+        self.assertEqual(st_5.size(), torch.Size([3, 4, 16]))
+        local_tensor = st_3.local_tensor()
+        dropout = torch.nn.Dropout(0.2)
+        local_tensor = torch.nn.functional.softmax(
+            local_tensor, dim=-1, dtype=torch.float32
+        )
+        st = torch.nn.functional.softmax(st_3, dim=-1, dtype=torch.float32)
+        self.assertEqual(st.local_tensor(), local_tensor)
+        torch.manual_seed(0)
+        local_tensor = dropout(local_tensor)
+        torch.manual_seed(0)
+        st_dropout = dropout(st)
+        self.assertEqual(st_dropout.local_tensor(), local_tensor)
```
For all of these ops (except dropout), we should do the following:

1. Create a local tensor for all ops.
2. Run these ops locally and get the result.
3. Shard the local tensor using (shard_tensor) and run the sharded tensor ops.
4. Verify local and sharded results are the same.

fduwjj replied:
Sure.. We definitely need to do that, initially this was to block the transformer use cases, so I didn't polish them that much. Let me make the test cases more robust and add more logics to handle corner cases.

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor_matrix_ops.py:
```diff
@@ -0,0 +1,108 @@
+    sys.exit(0)
+
+
+class TestShardedTernsorMatrixOps(ShardedTensorTestBase):
```
Why do we have two similar classes and test files? There is `test_sharded_tensor_matrix_ops.py` and also `test_matrix_ops.py`?

fduwjj replied:
Oh because one is for the class methods of sharded tensor and another one is for the ops.. If we go with the proposal above, I think we can consolidate these two test files.

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor_matrix_ops.py:
```diff
@@ -0,0 +1,108 @@
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_sharded_tensor_transpose(self):
+        specs = _chunk_sharding_specs_list_for_test([0], seed=7)
+        for spec in specs:
+            st = sharded_tensor.rand(spec, 10, 20, 5, init_rrefs=True)
+            self.assertEqual(3, st.dim())
+            self.assertEqual(torch.Size([10, 20, 5]), st.size())
+            local_tensor = st.local_tensor()
+            st = st.transpose(1, 0)
+            spec.dim = 1
+            local_tensor = local_tensor.transpose(1, 0)
+            st_2 = sharded_tensor.rand(spec, 20, 10, 5, init_rrefs=True)
+            self.assertEqual(st_2.size(), st.size())
+            self.assertEqual(local_tensor, st.local_tensor())
+            self.assertEqual(st_2.metadata(), st.metadata())
```
We are not really testing local and sharded version of the ops are same here since we are pulling out st.local_tensor(). We should run the op with a local tensor, shard the same local tensor and run the op and verify results.

fduwjj replied:
sure. Let me add that test cases.

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor_matrix_ops.py:
```diff
@@ -0,0 +1,108 @@
+        specs = _chunk_sharding_specs_list_for_test([0], seed=7)
+        for spec in specs:
+            st = sharded_tensor.rand(spec, 10, 22, 5, init_rrefs=True)
+            st = st.transpose(1, 0)
```
assert that sharded tensor is not contiguous here.

fduwjj replied:
This is tricky, because st.transpose cannot be not contiguous because when building a new st, we have to make it contiguous first...

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor_matrix_ops.py:
```diff
@@ -0,0 +1,108 @@
+    def test_sharded_tensor_masked_fill(self):
+        specs = _chunk_sharding_specs_list_for_test([0], seed=7)
+        for spec in specs:
+            st = sharded_tensor.rand(spec, 16, 30, 5, init_rrefs=True)
```
same as above, run local version and verify.

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor_matrix_ops.py:
```diff
@@ -0,0 +1,108 @@
+    @requires_nccl()
+    def test_sharded_tensor_view(self):
+        specs = _chunk_sharding_specs_list_for_test([0], seed=7)
+        for spec in specs:
```
same as above, run local version and verify.

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/chunk.py:
```diff
@@ -0,0 +1,35 @@
+        results.append(
+            ShardedTensor._init_from_local_tensor(
+                tensor.contiguous(),
+                input.sharding_spec(),
```
Sure.

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/chunk.py:
```diff
@@ -0,0 +1,35 @@
+    st_size = list(input.size())  # type: ignore[arg-type]
+    local_tensor = input.local_tensor()
+    tensor_chunks = torch.chunk(local_tensor, chunks, dim)
+    st_size[dim] //= chunks
```
Right, can we throw exception for now? Because the transformer code concatenate 3 matrixes together (q,k,v) so it must be divisible.

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/chunk.py:
```diff
@@ -0,0 +1,35 @@
+    tensor_chunks = torch.chunk(local_tensor, chunks, dim)
+    st_size[dim] //= chunks
+    results = []
+    for tensor in tensor_chunks:
```
Because inside transformer code, they concatenate k, v, q into one big matrix. But after chunk, it's actually three separate matrix. What I am thinking here is that:
1. If the chunk dim = sharding dim, the current behavior is to first split local tensor and then build a new st on top of. I guess what you are saying is that we want to first split the original tensor and then shard it again, right?
2. If the chunk dim != sharding dim, Then this logic is mathematically equivalent because the current way: to first split local tensor and then create a st is same as to first split the big tensor and then shard it again.

Is my understanding sound make sense to you?

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -0,0 +1,76 @@
+        raise TypeError("input or mat2 needs to be a ShardedTensor")
+    if input.dim() != 3 or mat2.dim() != 3:
+        raise TypeError("input or mat2 needs to be a 3D ShardedTensor")
+    local_tensor = torch.bmm(input.local_tensor(), mat2.local_tensor())
+    new_st_size = list(input.size())  # type: ignore[arg-type]
+    new_sharding_dim = input.sharding_spec().dim  # type: ignore[attr-defined]
+    mat2_sharding_dim = mat2.sharding_spec().dim  # type: ignore[attr-defined]
```
Not necessary, in the transformer code base, there are cases when both tensor are sharded in the last dim (-1)

fduwjj replied:
Resolved offline.

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -0,0 +1,76 @@
+    new_sharding_dim = input.sharding_spec().dim  # type: ignore[attr-defined]
+    mat2_sharding_dim = mat2.sharding_spec().dim  # type: ignore[attr-defined]
+    new_st_size[-1] = mat2.size()[-1]  # type: ignore[index]
+    if new_sharding_dim == -1 or new_sharding_dim == (input.dim() - 1):
+        if mat2_sharding_dim != -1 and mat2_sharding_dim != (mat2.dim() - 1):
+            new_st_size[-1] *= dist.get_world_size(pg)
```
Sure, so the reason for this logic is that if the input is sharded by -1 and the mat2 is sharded in other dimensions (for example, 0 or 1). If we will want to shard it by -1, we then need to multiply the world_size because new_st_size[-1] now is the local tensor size.

Of course, there are many corner cases we need to add down the road.

fduwjj replied:
Resolved offline.

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -0,0 +1,76 @@
+    if not isinstance(input, ShardedTensor):
+        raise TypeError("input needs to be a ShardedTensor")
+    local_tensor = input.local_tensor()
+    local_tensor = torch.nn.functional.softmax(local_tensor, dim=dim, dtype=dtype)
```
Right, your suggestion here absolutely makes sense. But this will add one more communication here. That's why initially I am not sure if we want it or not.

fduwjj commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -684,6 +688,169 @@ def local_tensor(self) -> torch.Tensor:
+            init_rrefs=self._init_rrefs,
+        )
+
+    def masked_fill(self, mask, value) -> ShardedTensor:
```
Looks like the dispath does not capture that. So if user call st.masked_fill and torch.masked_fill. The code goes down a different path. And I think keeping this class small really makes sense. So @wanchaol feel free to share your thoughts here.

wanchaol replied:
I think we will need to make `ShardedTensor` a `torch.Tensor` subclass in order for `st.masked_fill` to go through the same path as `torch.masked_fill`, maybe you can try patching https://github.com/pytorch/pytorch/pull/74695, In the meanwhile, I'm trying to land this PR asap, need to address a few comments, will update it soon today

fduwjj commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -684,6 +688,169 @@ def local_tensor(self) -> torch.Tensor:
             raise NotImplementedError("Only single local shard is supported.")
         return self.local_shards()[0].tensor
 
+    def contiguous(self) -> ShardedTensor:
```
Unfortunately, nope...

fduwjj commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -684,6 +688,169 @@ def local_tensor(self) -> torch.Tensor:
+        total_size = math.prod(st_size)  # type: ignore[attr-defined]
+        new_st_size = list(shape)
+        new_total_size = math.prod(new_st_size)  # type: ignore[attr-defined]
+        if (new_total_size < 0):
```
I mean you can only have one `-1` right? How can you infer the dimensions if you have two `-1`s? Or am I missing something here?

fduwjj commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -684,6 +688,169 @@ def local_tensor(self) -> torch.Tensor:
+            init_rrefs=self._init_rrefs,
+        )
+
+    def view(self, *shape) -> ShardedTensor:
+        """
+        Returns a new sharded tensor with the same data as the
+        self tensor but of a different shape for its local tensor.
+
+        For now, we only support to pass through the view op to the local
+        tensor.
```
Totally, the whole `view` api has loads of corner cases and we might need to reshard the tensor if the shape changes the dimensions of the shardedtensor. There could be more corner cases we need to consider. So theoretically, it should be equivalent to big tensor call the view first and shard the tensor, right?

fduwjj commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -684,6 +688,169 @@ def local_tensor(self) -> torch.Tensor:
+        return ShardedTensor._init_from_local_tensor(
+            new_local_tensor.contiguous(),
+            self.sharding_spec(),
+            st_size,
```
Not really, `new_st_size` is just the local tensor size not the global st tensor size. I should have made the naming more clearer. Let me change it.

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of TextRay (which is essentially XLMR model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of FairSeq and MetaSeq codebase (which is essentially a transformer model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example:

Ops include: chuck, transpose, view, mask_fill, dropout, softmax and type_as.

Isolate the common logic of registering simple ops into a function and for future register, we just need to implement at most three functions for a new op.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/chunk.py:
```diff
@@ -1,35 +1,235 @@
+        chunk_split_size = [len(chunk) for chunk in chunks]
+        local_tensor_chunks = torch.chunk(local_tensor, chunk_num, dim)
+
+        # When performing chunk on the dim of sharding, we need to shuffle the tensor.
```
So this is actually more complex than I though it is for the following reasons. Let's take the case of a regular torch.chunk on 1d Tensor of size 100. If we create 4 chunks, then it returns 4 tensors, where tensor 1 is 0-24, tensor 2 is 25-49 and so on. Now the same semantics should hold for ShardedTensor.

Now imagine a ShardedTensor again of size 100 and 1d but sharded across for ranks. So rank 1 has 0-24, rank 2 25-49 and so on.

Now, we if we call torch.chunk on this tensor with num_chunks=4, then it should create 4 ShardedTensors where tensor 1 is 0-24, tensor 2 is 25-49 and so on.

Now this means that ShardedTensor1 only exists on rank0, ShardedTensor2 only on rank 1 and so on. So the sharding spec has actually changed.

I feel for now it is probably best to error out in this case and only allow chunk on the sharded dimension. I'm assuming that is what we need for Transformers right?

fduwjj replied:
> Now this means that ShardedTensor1 only exists on rank0, ShardedTensor2 only on rank 1 and so on. So the sharding spec has actually changed.

But this is not the expected behavior right? 

We just first chunk them into 4 parts and do the sharding again. 

For transformers, we can error out it but I guess since we have already done it, I just include it in the PR.

fduwjj replied:
Discussed offline, we will make this error out for now.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/_common.py:
```diff
@@ -14,6 +21,171 @@
+        )
+
+
+def _sharded_op_common(op, early_stop_func, extra_check):
```
These are great utility functions!

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:
```diff
@@ -120,3 +126,70 @@ def binary_math_op(types, args=(), kwargs=None, pg=None):
+    """
+    if len(args) < 2:
+        raise TypeError("Needs two tensors to perform torch.bmm.")
+    st = args[0]
+    mat2 = args[1]
```
nit: Maybe st1 and st2 or mat1 and mat2 to be consistent?

fduwjj replied:
sure.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -1,76 +1,328 @@
+    if (
+        shape_size > st_size
+        or st_size % shape_size
+        or shapes.count(lambda x: x < -1) > 0
```
Shouldn't this be <=1?

fduwjj replied:
Oh, one -1 is allowed, but more than one -1 is invalid.

fduwjj replied:
After a second thought on this, will simply this logic a little bit.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -1,76 +1,328 @@
+    st = args[0]
+    shapes = args[1:]
+    # Infer the dim which is specified with -1.
+    if shapes.count(-1):
```
nit: Maybe just use `shapes.index(-1)` here and catch and ignore the exception instead of first using count then index.

fduwjj replied:
Ah.. I see. Let me change it.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -1,76 +1,328 @@
+            if placement.rank() == current_rank:  # type: ignore[index]
+                rank_idx = idx  # type: ignore[attr-defined]
+        shard_metadata = st.metadata().shards_metadata[rank_idx]  # type: ignore[index]
+        mask = mask.narrow(
+            narrow_idx,
+            shard_metadata.shard_offsets[sharding_dim],
+            shard_metadata.shard_sizes[sharding_dim],
+        )
```
So this is assuming the mask is a ReplicatedTensor which is the same on all ranks and the value is a regular torch.Tensor which is different on each rank?

In the local version of this op, the original tensor, mask and value are of same size. But here it seems a bit different? Maybe mask needs to be a ReplicatedTensor and value needs to be a ShardedTensor?

In our particular Transformer use case, how are these tensors set up?

fduwjj replied:
They are same across all ranks. We just need to narrow them to the part which is stored on the current rank.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -1,76 +1,328 @@
+def sharded_transpose_check(*args, **kwargs):
+    """
+    Perform extra checks for the sharded_transpose op such as the input needs to
+    be at least 3 and the sharding spec needs to be a ChunkShardingSpec.
```
at least 2?

fduwjj replied:
Ahh, yes.. Might be a typo here.

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/chunk.py:
```diff
@@ -1,35 +1,235 @@
+        sharding_dim = spec.dim
+        local_tensor = st.local_tensor()
+        chunks = torch.arange(st_size[dim]).chunk(chunk_num)
+        chunk_split_size = [len(chunk) for chunk in chunks]
+        local_tensor_chunks = torch.chunk(local_tensor, chunk_num, dim)
```
Oh I see.. Let me try it.

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of FairSeq and MetaSeq codebase (which is essentially a transformer model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example:

Ops include: chuck, transpose, view, mask_fill, dropout, softmax and type_as.

Isolate the common logic of registering simple ops into a function and for future register, we just need to implement at most three functions for a new op.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of FairSeq and MetaSeq codebase (which is essentially a transformer model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example:

Ops include: chuck, transpose, view, mask_fill, dropout, softmax and type_as.

Isolate the common logic of registering simple ops into a function and for future register, we just need to implement at most three functions for a new op.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

fduwjj commented:
Addressed comments from reviewers and fix linter errors.

wanchaol commented on torch/distributed/_shard/sharded_tensor/_ops/chunk.py:
```diff
@@ -0,0 +1,63 @@
+
+def register_chunk_op(op):
+    @sharded_op_impl(op)
+    def sharded_chunk(types, args=(), kwargs=None, pg=None):
```
why the `chunk` op could't use `_register_sharded_op_on_local_tensor`? I feel we could use this and just need to add `extra_check` condition to ensure it's not chunking the sharding dim?

fduwjj replied:
The return value here is different. Chunk return a list of sharded tensors.

wanchaol commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -0,0 +1,330 @@
+
+def sharded_type_as(args, kwargs, pg):
+    """
+    Handles ``__torch_function__`` dispatch for the sharded_type_as op.
```
nit: for the `Tensor.type_as` op

wanchaol commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -0,0 +1,330 @@
+        _swap_meta_data(shard_meta_data.shard_offsets, dim0, dim1)
+        _swap_meta_data(shard_meta_data.shard_sizes, dim0, dim1)
+        new_local_shards.append(
+            Shard(shard.tensor.transpose(dim0, dim1).contiguous(), shard_meta_data)
```
should we insert the `contiguous()` call implicitly? or we rely on user to call `contiguous()` after the `transpose`? I feel like we shouldn't do it implicitly here unless we have a reason to

fduwjj replied:
I agree but Shard here will throw exception if we don't call `contiguous` here..

wanchaol commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -0,0 +1,330 @@
+            Shard(shard.tensor.transpose(dim0, dim1).contiguous(), shard_meta_data)
+        )
+    st_meta = copy.deepcopy(st.metadata())
+    for shard_metadata in st_meta.shards_metadata:
```
nit: instead of swapping it again, we should just construct a new `ShardedTensorMetadata`, with tensor size swapped, tensor properties copied from st.metadata(), set `shards_metadata` with the swapped ones we recorded earlier.

fduwjj replied:
But previous ones only have the local shard metadata right? After a second, I change it to only work for a local tensor for now. There might be more corner cases when it comes to more than one local shard per node.

wanchaol replied:
if this is the case, should we throw if there's more multiple local shards in one rank?

wanchaol replied:
`st.metadata()` should contains size, tensor properties info, except that the shards metadata should be the swapped one. My point here is since you already swapped the shard_metadata when constructing the local shards, we should directly use the constructed local shard metadata list instead of deep copy a new one and swap them again? It's not a big deal and we could be enhanced in a separate PR, but I think the doing things again instead of a single source of truth might be an issue in the future.

wanchaol commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -0,0 +1,330 @@
+    shape_size = math.prod(shapes)  # type: ignore[attr-defined]
+    neg_sum = sum(i for i in shapes if i < 0)
+    if shape_size > st_size or st_size % shape_size:
+        raise ValueError("Shape is invalid for sharded tensor size.")
```
maybe let's throw the same error as the torch.Tensor case?

wanchaol reviewed (commented):
Looks great! Some suggestions on simplifying decorator and some nits inlined.

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/_common.py:
```diff
@@ -14,6 +21,171 @@
+
+
+def _sharded_op_common(op, early_stop_func, extra_check):
+    """
```
Originally, I thought `sharded_op_impl` is just for op registration, so I want to decouple the logic here. Maybe let's land this for ops enabling first. I think I can send a follow-up PR to make it cleaner and @pritamdamania87 is also trying to refactor the structure.

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/_common.py:
```diff
@@ -14,6 +21,171 @@
+                raise ValueError(f" No input for '{op.__name__}'!")
+            # Validate types
+            st = args[0]
+            if not isinstance(st, ShardedTensor):
```
I would image the first param is always a sharded tensor? Do you have a detailed use case example?

wanchaol replied:
hmm I think there's bunch of common use cases where args[0] is not sharded tensor that we need to consider, i.e. for a operation like `3 + st`(it will have args with the first one be `3` and the second one be `st), `replicated_tensor + st`, `, then args[0] won't be a ShardedTensor?

wanchaol replied:
Yeah i guess if this decorator is only used for the functions that registered in this PR, I think it make sense to separate them for now, we might need to consider the cases in the future.

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py:
```diff
@@ -0,0 +1,330 @@
+    Args: same as ``torch.Tensor.type_as``.
+
+    Return (bool): Whether to return early or not.
+    """
```
Do you mean the case like `tensor.type_as(sharded_tensor)` ? Right now, we have not seen such use cases, that's why we enforce args[0] to be a sharded tensor.

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/ops/test_math_ops.py:
```diff
@@ -16,19 +15,17 @@
 class TestMathOps(ShardedTensorTestBase):
     @with_comms(init_rpc=False)
-    @skip_if_lt_x_gpu(TEST_GPU_NUM)
+    @skip_if_lt_x_gpu(4)
```
Why change this to 4? Isn't `TEST_GPU_NUM` more generic?

fduwjj replied:
Opps, I think I use `TEST_GPU_NUM ` originally. Let me change it.

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of FairSeq and MetaSeq codebase (which is essentially a transformer model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example:

Ops include: chuck, transpose, view, mask_fill, dropout, softmax and type_as.

Isolate the common logic of registering simple ops into a function and for future register, we just need to implement at most three functions for a new op.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of FairSeq and MetaSeq codebase (which is essentially a transformer model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example:

Ops include: chuck, transpose, view, mask_fill, dropout, softmax and type_as.

Isolate the common logic of registering simple ops into a function and for future register, we just need to implement at most three functions for a new op.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of FairSeq and MetaSeq codebase (which is essentially a transformer model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example:

Ops include: chuck, transpose, view, mask_fill, dropout, softmax and type_as.

Isolate the common logic of registering simple ops into a function and for future register, we just need to implement at most three functions for a new op.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Enable ops needed in the transformer model training"


From the code base of FairSeq and MetaSeq codebase (which is essentially a transformer model), we have found that loads of ops are not supported by sharded tensor. So we now implement a simple version so that we can at least run a transformer example:

Ops include: chuck, transpose, view, mask_fill, dropout, softmax and type_as.

Isolate the common logic of registering simple ops into a function and for future register, we just need to implement at most three functions for a new op.

Differential Revision: [D35123021](https://our.internmc.facebook.com/intern/diff/D35123021/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35123021/)!

[ghstack-poisoned]
